### PR TITLE
Js impl of refine accepted length for ph extensions pr#2526

### PR DESCRIPTION
**Pull request by penmetsaa**
Please refer PR #2526 for details about this change.

### CODE REVIEW

File: javascript/i18n/phonenumbers/phonenumberutil.js [hunk 1]
```diff
@@ -743,47 +743,122 @@ i18n.phonenumbers.PhoneNumberUtil.VALID_PHONE_NUMBER_ =
 i18n.phonenumbers.PhoneNumberUtil.DEFAULT_EXTN_PREFIX_ = ' ext. ';
 
 
-/**
- * Pattern to capture digits used in an extension.
- * Places a maximum length of '7' for an extension.
+/**  Helper method for constructing regular expressions for parsing. Creates
```
Comment (lararennie) [inline javascript/i18n/phonenumbers/phonenumberutil.js:746]: Start comments on the next line after /**

File: javascript/i18n/phonenumbers/phonenumberutil.js [hunk 2]
```diff
@@ -743,47 +743,122 @@ i18n.phonenumbers.PhoneNumberUtil.VALID_PHONE_NUMBER_ =
...
+ *   an expression that captures up to max_length digits.
  *
- * @const
- * @type {string}
- * @private
+ *   @return {string} constructued RegEx pattern to capture extension digits.
```
Comment (lararennie) [inline javascript/i18n/phonenumbers/phonenumberutil.js:749]: remove "constructued" (typo anyway and doesn't add clarity)

File: javascript/i18n/phonenumbers/phonenumberutil.js [hunk 3]
```diff
@@ -743,47 +743,122 @@ i18n.phonenumbers.PhoneNumberUtil.VALID_PHONE_NUMBER_ =
...
- * the digits that you want to capture as part of the extension, or else parsing
- * will fail! We allow two options for representing the accented o - the
- * character itself, and one in the unicode decomposed form with the combining
- * acute accent.
+ * Helper initialiser method to create the regular-expression pattern to match
+ * extensions.
  *
- * @const
- * @type {string}
+ * @return {string} constructued RegEx pattern to capture extensions.
```
Comment (lararennie) [inline javascript/i18n/phonenumbers/phonenumberutil.js:762]: same here

File: javascript/i18n/phonenumbers/phonenumberutil.js [hunk 4]
```diff
@@ -743,47 +743,122 @@ i18n.phonenumbers.PhoneNumberUtil.VALID_PHONE_NUMBER_ =
...
+ var ambiguousExtLabels = "(?:[x\uFF58#\uFF03~\uFF5E]|int|\uFF49\uFF4E\uFF54)";
+ // When extension is not separated clearly.
+ /** @type {string} */
+ var ambiguousSeparator = "[- ]+";
+ // ",," is commonly used for auto dialling the extension when connected. First
+ // comma is matched through possibleSeparatorsBetweenNumberAndExtLabel, so we do
+ // not repeat it here. Semi-colon works in Iphone and Android also to pop up a
+ // button with the extension number following.
+ /** @type {string} */
+ var autoDiallingAndExtLabelsFound = "(?:,{2}|;)";
```
Comment (lararennie) [inline javascript/i18n/phonenumbers/phonenumberutil.js:803]: Move to below possibleSeparatorsNumberExtLabelNoComma for consistency with Java

Comment (lararennie) [reply javascript/i18n/phonenumbers/phonenumberutil.js:803]: Actually in general can you reorder these definitions to match Java?

File: javascript/i18n/phonenumbers/phonenumberutil.js [hunk 5]
```diff
@@ -743,47 +743,122 @@ i18n.phonenumbers.PhoneNumberUtil.VALID_PHONE_NUMBER_ =
...
+ // This is same as possibleSeparatorsBetweenNumberAndExtLabel, but not matching
+ // comma as extension label may have it.
+ /** @type {string} */
+ var possibleSeparatorsNumberExtLabelNoComma = "[ \u00A0\\t]*";
+
+ /** @type {string} */
+ var autoDiallingExtn = possibleSeparatorsNumberExtLabelNoComma
+        + autoDiallingAndExtLabelsFound + possibleCharsAfterExtLabel
+        + i18n.phonenumbers.PhoneNumberUtil.extnDigits_(extLimitAfterLikelyLabel)
+	+ optionalExtnSuffix;
```
Comment (lararennie) [inline javascript/i18n/phonenumbers/phonenumberutil.js:832]: this + looks misaligned.

File: javascript/i18n/phonenumbers/phonenumberutil.js [hunk 6]
```diff
@@ -743,47 +743,122 @@ i18n.phonenumbers.PhoneNumberUtil.VALID_PHONE_NUMBER_ =
...
+ /** @type {string} */
+ var extensionPattern =
+     rfcExtn + "|"
+     + explicitExtn + "|"
+     + ambiguousExtn + "|"
+     + americanStyleExtnWithSuffix + "|"
+     + autoDiallingExtn + "|"
+     + onlyCommasExtn;
+
+ return extensionPattern;
```
Comment (lararennie) [inline javascript/i18n/phonenumbers/phonenumberutil.js:860]: I wouldn't bother constructing it separately, you can just do return rfcExtn + ....

File: javascript/i18n/phonenumbers/phonenumberutil_test.js [hunk 7]
```diff
@@ -3550,6 +3550,151 @@ function testParseExtensions() {
...
 }
 
+function testParseHandlesLongExtensionsWithExplicitLabels() {
+  // Test lower and upper limits of extension lengths for each type of label.
+  /** @type {!i18n.phonenumbers.PhoneNumber} */
+  var nzNumber = new i18n.phonenumbers.PhoneNumber();
+  nzNumber.setCountryCode(64);
+  nzNumber.setNationalNumber(33316005);
+
+  // Firstly, when in RFC format: PhoneNumberUtil.CAPTURING_LONG_EXTN_DIGITS_WHEN_CLEAR
```
Comment (lararennie) [inline javascript/i18n/phonenumbers/phonenumberutil_test.js:3560]: These CAPTURING_LONG_EXTN_DIGITS_WHEN_CLEAR bits etc need updating to refer to vars that you actually declare

Comment (lararennie) [reply javascript/i18n/phonenumbers/phonenumberutil_test.js:3560]: Same in other test cases below.

**Review by lararennie — COMMENTED** [review]
Have you run closure lint etc on it?

File: javascript/i18n/phonenumbers/phonenumberutil.js (hunk 1 above)
Comment (penmetsaa) [reply javascript/i18n/phonenumbers/phonenumberutil.js:746]: Done.

File: javascript/i18n/phonenumbers/phonenumberutil.js (hunk 2 above)
Comment (penmetsaa) [reply javascript/i18n/phonenumbers/phonenumberutil.js:749]: Done.

File: javascript/i18n/phonenumbers/phonenumberutil.js (hunk 3 above)
Comment (penmetsaa) [reply javascript/i18n/phonenumbers/phonenumberutil.js:762]: Done.

File: javascript/i18n/phonenumbers/phonenumberutil.js (hunk 4 above)
Comment (penmetsaa) [reply javascript/i18n/phonenumbers/phonenumberutil.js:803]: Done.

File: javascript/i18n/phonenumbers/phonenumberutil.js (hunk 5 above)
Comment (penmetsaa) [reply javascript/i18n/phonenumbers/phonenumberutil.js:832]: Not able to view the issue. Maybe I have already corrected.

File: javascript/i18n/phonenumbers/phonenumberutil.js (hunk 6 above)
Comment (penmetsaa) [reply javascript/i18n/phonenumbers/phonenumberutil.js:860]: Done.

File: javascript/i18n/phonenumbers/phonenumberutil_test.js (hunk 7 above)
Comment (penmetsaa) [reply javascript/i18n/phonenumbers/phonenumberutil_test.js:3560]: Oh yah. true. Some stale comments as I started JS some time back.
Corrected now.

File: javascript/i18n/phonenumbers/phonenumberutil.js (hunk 5 above)
Comment (penmetsaa) [reply javascript/i18n/phonenumbers/phonenumberutil.js:834]: In view of actual completed file, it is aligned rightly. :(

File: javascript/i18n/phonenumbers/phonenumberutil.js [hunk 8]
```diff
@@ -795,14 +799,16 @@ i18n.phonenumbers.PhoneNumberUtil.createExtnPattern_ =
  // When extension is not separated clearly.
  /** @type {string} */
  var ambiguousSeparator = "[- ]+";
+ // This is same as possibleSeparatorsBetweenNumberAndExtLabel, but not matching
```
Comment (lararennie) [inline javascript/i18n/phonenumbers/phonenumberutil.js:802]: "the" before "same"

**Review by lararennie — DISMISSED** [review]

File: javascript/i18n/phonenumbers/phonenumberutil.js [hunk 9]
```diff
@@ -836,26 +837,25 @@ i18n.phonenumbers.PhoneNumberUtil.createExtnPattern_ =
...
+ // The first regular expression covers RFC 3966 format, where the extension is added
+ // using ";ext=". The second more generic where extension is mentioned with explicit
+ // labels like "ext:". In both the above cases we allow more numbers in extension than
+ // any other extension labels. The third one captures when single character extension
  // labels or less commonly used labels are used. In such cases we capture fewer
  // extension digits in order to reduce the chance of falsely interpreting two
  // numbers beside each other as a number + extension. The fourth one covers the
  // special case of American numbers where the extension is written with a hash
- // at the end, such as "- 503#". The fifth one exclusive for extension autodialling
+ // at the end, such as "- 503#". The fifth one exclusively for extension autodialling
```
[the remark this answers is not in the document]
Comment (penmetsaa) [reply javascript/i18n/phonenumbers/phonenumberutil.js]: Done.

File: javascript/i18n/phonenumbers/phonenumberutil.js (hunk 8 above)
Comment (penmetsaa) [reply javascript/i18n/phonenumbers/phonenumberutil.js]: Done.

**Review by lararennie — APPROVED** [review]

**Pull request merged**